opencog/util/Logger.cc

```diff
@@ -147,7 +148,9 @@ Logger::~Logger()
     flush();
     stop_write_loop();
 
+    std::unique_lock<std::mutex> lock(the_mutex);
     if (logfile != NULL) fclose(logfile);
+	 lock.unlock();   
```

Comment:
Weird indent

Reply:
My opinion it is better to replace ```lock.unlock()``` by destructor call to make the scope of the lock more explicit. For instance replace three lines of code above by:

    {
        std::unique_lock<std::mutex> lock(the_mutex);
        if (logfile != NULL) fclose(logfile);
    }

Reply:
yes, the_mutex is part of class Logger
If I understand your comment, you don't think is needed to unlock as the Logger object destroys the mutex anyway.

I was concerned about this:
The behavior of a program is undefined if a mutex is destroyed while still owned by any threads, or a thread terminates while owning a mutex. 
from std::mutex manual.

Also, investigating if need for a try_lock logic.
Looking into this as well:
std::mutex is usually not accessed directly: std::unique_lock, std::lock_guard, or std::scoped_lock (since C++17) manage locking in a more exception-safe manner.

Reply:
The mutex and the lock are not the same thing.  The mutex is in the class, and it is not destroyed until the class is destroyed. The mutex is shared by all threads accessing the calss.

The lock is a different C++ structure, and it is held only block scope:  it is created in the block, and destroyed when the block goes out of scope.  The lock is not shared between different threads: it just sits in the stack frame of the current subroutine call.

There is no need for try_lock, here. Try-locks are for solving various "thundering herd" and deadlock-related problems.  You don't have that here.

Reply:
Anyway, Vitaly knows this stuff. I am bowing out of this conversation, I've got too many other things to worry about.

Reply:
Hi @mselseth, my comment is more about code style than about code correctness. As far as I can see your code is correct but explicit ```lock.unlock()``` is just not needed when ```unique_lock``` is used properly.

As @linas described above ```the_mutex``` is shared between threads and it is not destroyed while instance of Logger is live. ```std::unique_lock<std::mutex> lock``` is a helper which locks a mutex in constructor and unlocks it in destructor. The question is "why is this helper needed?" because you always can use ```mutex::lock``` and ```mutex::unlock``` methods explicitly. The answer is: by using ```unique_lock``` you can easily release the mutex correctly even when exception was thrown between ```mutex::lock``` and ```mutex::unlock``` calls (for instance in ```fclose``` call above).

To use this pattern one should define an instance of ```unique_lock``` at the first line of the code block which should be executed while mutex is locked. ```unique_lock``` locks the mutex in the constructor. And when control flow leaves the code block (it doesn't matter whether it is an end of the block reached or exception thrown) ```unique_lock``` destructor is called and mutex is released.

So taking this into account:
- if it is necessary to hold mutex only for the part of the method code (like in code snippet above) it is safer to create new code block and define ```unique_lock``` at the first line. Otherwise further code changes may lead to the situations when mutex is not released until the end of the method. It will not necessary happen but it is just a bit safer way of doing things.
- if mutex should be locked while whole method is executed it is enough to define ```unique_lock``` at the first line of the method. It is not necessary call ```unique_lock::unlock``` at the end of the code block (see my comment below https://github.com/opencog/cogutil/pull/127#discussion_r213958098)